### ISSUE TITLE
Data races detected with `log`

### ISSUE THREAD

**Issue by ryepup** [Azure/azure-sdk-for-go#17369]
### Bug Report

- import path of package in question: `github.com/Azure/azure-sdk-for-go/sdk/azcore/log`
- SDK version: azcore@v0.22.0
- output of `go version`: `go1.18 darwin/amd64`

- What happened?

I have many tests running in parallel that call `log.SetListener` and `log.SetEvents` before using other azure-sdk-for-go features. I run my tests with `-race` enabled.

Occasionally I'll get a data race detected with `log.Should`, as multiple goroutines read and write to the process wide logger at 

https://github.com/Azure/azure-sdk-for-go/blob/3632133b2c1f953043f372494fac38744a93fd4c/sdk/internal/log/log.go#L97

The test output looks something like this:

```text
WARNING: DATA RACE
Read at 0x00c0008a2780 by goroutine 15:
  github.com/Azure/azure-sdk-for-go/sdk/internal/log.Should()
      ~/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/internal@v0.9.1/log/log.go:65 +0x16b
  github.com/Azure/azure-sdk-for-go/sdk/internal/log.Writef()
      ~/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/internal@v0.9.1/log/log.go:85 +0x64
  github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime.(*retryPolicy).Do()
      ~/go/pkg/mod/github.com/!azure/azure-sdk-for-go/sdk/azcore@v0.22.0/runtime/policy_retry.go:105 +0x712
  github.com/Azure/azure-sdk-for-go/sdk/azcore/internal/pipeline.(*Request).Next()
...
Previous write at 0x00c0008a2780 by goroutine 18:
  .../Client.EnableVerboseLogging()
      ~/Documents/projects/.../client.go:75 +0x47d
```

- What did you expect or want to happen?

No data races.

- How can we reproduce it?

Run many goroutines that call  `log.SetListener`, `log.SetEvents`, and `log.Should`.

Maybe guard that package var with a `sync.RWLock`? Or some docs on `SetEvents` and `SetListener` specifying they are not safe for concurrent usage?

**Comment by jhendrixMSFT** [comment]
Would you mind sharing a little bit about your scenario?  When we wrote this, we envisioned that this is something that would be called by a program once during start-up and would remain static thereafter.  I'd like to hear about how you use this in case we need to reconsider.

**Comment by ryepup** [comment]
My scenario is `go test -race ...` with integration tests that look roughly like:

```
func newSB() azservicebus.Client {
   azlog.SetEvents(...)
   azlog.SetListener(...)
   // create and return the client
}

func TestOne(t *testing.T) {
  t.Parallel()
  c := newSB()
  // do some operations
}

func TestTwo(t *testing.T) {
  t.Parallel()
  c := newSB()
  // do some operations
}
```

> When we wrote this, we envisioned that this is something that would be called by a program once during start-up and would remain static thereafter.

That might be nice to add to the doc comments on `SetEvents` and `SetListener`. I could imagine someone implementing a feature to dynamically change log levels to support debugging live systems.

I understand this is not the expected usage patterns, but figured you might want to know about the race. Feel free to close this out as a "wontfix", I solved my problem by changing my tests.

**Comment by jhendrixMSFT** [comment]
Thanks for the info.  I'll discuss with the team how we want to proceed.  At the very least, I agree that we should at least update the docs indicating this isn't goroutine-safe if we leave it as-is.
